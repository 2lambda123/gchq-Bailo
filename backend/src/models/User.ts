// This interface stores information about the properties on the base object.
// It should be used for plain object representations, e.g. for sending to the
// client.
export interface UserInterface {
  // Do not store user role information on this object.  This information
  // should be stored in an external corporate store.
  dn: string
}
