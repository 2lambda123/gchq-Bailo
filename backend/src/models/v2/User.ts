import { Document, model, Schema } from 'mongoose'

// This interface stores information about the properties on the base object.
// It should be used for plain object representations, e.g. for sending to the
// client.
export interface UserInterface {
  // Do not store user role information on this object.  This information
  // should be stored in an external corporate store.
  dn: string

  createdAt: Date
  updatedAt: Date
}

// The doc type includes all values in the plain interface, as well as all the
// properties and functions that Mongoose provides.  If a function takes in an
// object from Mongoose it should use this interface
export type UserDoc = UserInterface & Document<any, any, UserInterface>

const UserSchema = new Schema<UserInterface>(
  {
    dn: { type: String, required: true, unique: true, index: true },
  },
  {
    timestamps: true,
    collection: 'v2_users',
  }
)

const UserModel = model<UserInterface>('v2_User', UserSchema)

export default UserModel
