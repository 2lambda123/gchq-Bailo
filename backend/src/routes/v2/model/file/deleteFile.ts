import bodyParser from 'body-parser'
import { Request, Response } from 'express'
import { z } from 'zod'

import { removeFile } from '../../../../services/v2/file.js'
import { parse } from '../../../../utils/validate.js'

export const deleteFileSchema = z.object({
  params: z.object({
    modelId: z.string(),
    fileId: z.string(),
  }),
})

interface DeleteFileResponse {
  message: string
}

export const deleteFile = [
  bodyParser.json(),
  async (req: Request, res: Response<DeleteFileResponse>) => {
    const {
      params: { modelId, fileId },
    } = parse(req, deleteFileSchema)

    await removeFile(req.user, modelId, fileId)

    return res.json({
      message: 'Successfully removed file.',
    })
  },
]
