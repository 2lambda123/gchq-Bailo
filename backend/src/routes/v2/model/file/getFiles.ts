import bodyParser from 'body-parser'
import { Request, Response } from 'express'
import { z } from 'zod'

import { parse } from '../../../../utils/validate.js'
import { FileCategory, FileInterface } from '../../../../models/v2/File.js'

export const getFilesSchema = z.object({
  params: z.object({
    modelId: z.string({
      required_error: 'Must specify model id as param',
    }),
  }),
})

interface GetFilesResponse {
  data: {
    files: Array<FileInterface>
  }
}

export const getFiles = [
  bodyParser.json(),
  async (req: Request, res: Response<GetFilesResponse>) => {
    const _ = parse(req, getFilesSchema)

    return res.json({
      data: {
        files: [
          {
            modelId: 'example-model',

            name: 'example-file',
            category: FileCategory.Other,
            size: 1024,

            bucket: 'uploads',
            path: '/example/upload/path',

            complete: true,

            createdAt: new Date(),
            updatedAt: new Date(),
          },
        ],
      },
    })
  },
]
