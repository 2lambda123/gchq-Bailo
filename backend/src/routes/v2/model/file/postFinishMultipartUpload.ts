import { Request, Response } from 'express'
import { z } from 'zod'

import { parse } from '../../../../middleware/validate.js'

export const postFinishMultipartUploadSchema = z.object({
  params: z.object({
    modelId: z.string(),
  }),
  body: z.object({
    fileId: z.string(),
    parts: z.array(z.unknown()),
  }),
})

interface PostFinishMultipartUpload {
  data: {
    message: string
  }
}

export const postFinishMultipartUpload = [
  async (req: Request, res: Response<PostFinishMultipartUpload>) => {
    const _ = parse(req, postFinishMultipartUploadSchema)

    return res.json({
      data: {
        message: 'Successfully finished multipart upload.',
      },
    })
  },
]
