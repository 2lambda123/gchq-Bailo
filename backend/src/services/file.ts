import { putObjectStream } from '../clients/s3.js'
import authorisation, { ModelAction } from '../connectors/v2/authorisation/index.js'
import FileModel from '../models/v2/File.js'
import { UserDoc } from '../models/v2/User.js'
import config from '../utils/v2/config.js'
import { Forbidden } from '../utils/v2/error.js'
import { longId } from '../utils/v2/id.js'
import { getModelById } from './v2/model.js'

export async function uploadModelFile(
  user: UserDoc,
  modelId: string,
  name: string,
  mime: string,
  stream: ReadableStream
) {
  const model = await getModelById(user, modelId)

  if (!(await authorisation.userModelAction(user, model, ModelAction.UploadFile))) {
    throw Forbidden(`You do not have permission to upload a file to this model.`, { userDn: user.dn })
  }

  const fileId = longId()

  const bucket = config.s3.buckets.uploads
  const path = `beta/model/${modelId}/files/${fileId}`

  const { fileSize } = await putObjectStream(bucket, path, stream)

  const file = new FileModel({
    modelId,
    name,
    mime,
    bucket,
    path,
    size: fileSize,
    complete: true,
  })

  await file.save()

  return file
}
