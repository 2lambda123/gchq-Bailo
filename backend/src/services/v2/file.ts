import { getObjectStream, putObjectStream } from '../../clients/s3.js'
import { FileAction } from '../../connectors/v2/authorisation/Base.js'
import authorisation from '../../connectors/v2/authorisation/index.js'
import FileModel from '../../models/v2/File.js'
import { UserDoc } from '../../models/v2/User.js'
import { asyncFilter } from '../../utils/general.js'
import config from '../../utils/v2/config.js'
import { Forbidden, NotFound } from '../../utils/v2/error.js'
import { longId } from '../../utils/v2/id.js'
import { getModelById } from './model.js'
import { removeFileFromReleases } from './release.js'

export async function uploadFile(user: UserDoc, modelId: string, name: string, mime: string, stream: ReadableStream) {
  const model = await getModelById(user, modelId)

  const fileId = longId()

  const bucket = config.s3.buckets.uploads
  const path = `beta/model/${modelId}/files/${fileId}`

  const file = new FileModel({
    modelId,
    name,
    mime,
    bucket,
    path,
    complete: true,
  })

  if (!(await authorisation.userFileAction(user, model, file, FileAction.Upload))) {
    throw Forbidden(`You do not have permission to upload a file to this model.`, { userDn: user.dn })
  }

  const { fileSize } = await putObjectStream(bucket, path, stream)
  file.size = fileSize

  await file.save()

  return file
}

export async function downloadFile(user: UserDoc, fileId: string, range?: { start: number; end: number }) {
  const file = await getFileById(user, fileId)
  const model = await getModelById(user, file.modelId)

  const access = await authorisation.userFileAction(user, model, file, FileAction.Download)
  if (!access) {
    throw Forbidden(`You do not have permission to download this model.`, { user: user.dn, fileId })
  }

  return getObjectStream(file.bucket, file.path, range)
}

export async function getFileById(user: UserDoc, fileId: string) {
  const file = await FileModel.findOne({
    _id: fileId,
  })

  if (!file) {
    throw NotFound(`The requested file was not found.`, { fileId })
  }

  const model = await getModelById(user, file.modelId)

  if (!(await authorisation.userFileAction(user, model, file, FileAction.View))) {
    throw Forbidden(`You do not have permission to get this file.`, { userDn: user.dn, fileId })
  }

  return file
}

export async function getFilesByModel(user: UserDoc, modelId: string) {
  const model = await getModelById(user, modelId)
  const files = await FileModel.find({ modelId })

  return asyncFilter(files, (file) => authorisation.userFileAction(user, model, file, FileAction.View))
}

export async function getFilesByIds(user: UserDoc, modelId: string, fileIds: string[]) {
  const model = await getModelById(user, modelId)
  const files = await FileModel.find({ _id: { $in: fileIds } })

  if (files.length !== fileIds.length) {
    const notFoundFileIds = fileIds.filter((id) => files.some((file) => file._id.toString() === id))
    throw NotFound(`The requested files were not found.`, { fileIds: notFoundFileIds })
  }

  return asyncFilter(files, (file) => authorisation.userFileAction(user, model, file, FileAction.View))
}

export async function removeFile(user: UserDoc, modelId: string, fileId: string) {
  const model = await getModelById(user, modelId)
  const file = await getFileById(user, fileId)

  if (!(await authorisation.userFileAction(user, model, file, FileAction.Delete))) {
    throw Forbidden(`You do not have permission to delete a file from this model.`, { userDn: user.dn })
  }

  await removeFileFromReleases(user, model, fileId)

  // We don't actually remove the file from storage, we only hide all
  // references to it.  This makes the file not visible to the user.
  await file.delete()

  return file
}
