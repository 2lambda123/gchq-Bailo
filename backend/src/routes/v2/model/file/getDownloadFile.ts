import bodyParser from 'body-parser'
import contentDisposition from 'content-disposition'
import { Request, Response } from 'express'
import stream from 'stream'
import { z } from 'zod'

import { FileInterface } from '../../../../models/v2/File.js'
import { downloadFile, getFileById } from '../../../../services/v2/file.js'
import { BadReq, InternalError } from '../../../../utils/v2/error.js'
import { parse } from '../../../../utils/validate.js'

export const getDownloadFileSchema = z.object({
  params: z.object({
    modelId: z.string(),
    fileId: z.string(),
  }),
})

interface GetDownloadFileResponse {
  files: Array<FileInterface>
}

export const getDownloadFile = [
  bodyParser.json(),
  async (req: Request, res: Response<GetDownloadFileResponse>) => {
    const {
      params: { fileId },
    } = parse(req, getDownloadFileSchema)

    const file = await getFileById(req.user, fileId)

    // required to support utf-8 file names
    res.set('Content-Disposition', contentDisposition(file.name, { type: 'inline' }))
    res.set('Content-Type', file.mime)
    res.set('Cache-Control', 'public, max-age=604800, immutable')

    if (req.headers.range) {
      // TODO: support ranges
      throw BadReq('Ranges are not supported', { fileId })
    }

    res.set('Content-Length', String(file.size))
    // TODO: support ranges
    // res.set('Accept-Ranges', 'bytes')
    res.writeHead(200)

    const stream = await downloadFile(req.user, fileId)
    if (!stream.Body) {
      throw InternalError('We were not able to retrieve the body of this file', { fileId })
    }

    // The AWS library doesn't seem to properly type 'Body' as being pipeable?
    ;(stream.Body as stream.Readable).pipe(res)
  },
]
