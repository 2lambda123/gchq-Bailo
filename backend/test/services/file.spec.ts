import { Readable } from 'stream'
import { describe, expect, test, vi } from 'vitest'

import { UserDoc } from '../../src/models/v2/User.js'
import { getFilesByModel, removeFile, uploadFile } from '../../src/services/v2/file.js'

vi.mock('../../src/utils/config.js')

const s3Mocks = vi.hoisted(() => ({
  putObjectStream: vi.fn(() => ({ fileSize: 100 })),
}))
vi.mock('../../src/clients/s3.js', () => s3Mocks)

const authorisationMocks = vi.hoisted(() => ({
  userModelAction: vi.fn(() => true),
}))
vi.mock('../../src/connectors/v2/authorisation/index.js', async () => ({
  default: authorisationMocks,
}))

const modelMocks = vi.hoisted(() => ({
  getModelById: vi.fn(),
}))
vi.mock('../../src/services/v2/model.js', () => modelMocks)

const fileModelMocks = vi.hoisted(() => {
  const obj: any = {}

  obj.save = vi.fn(() => obj)
  obj.find = vi.fn(() => obj)
  obj.findOne = vi.fn(() => obj)
  obj.delete = vi.fn(() => obj)

  const model: any = vi.fn(() => obj)
  Object.assign(model, obj)

  return model
})
vi.mock('../../src/models/v2/File.js', () => ({ default: fileModelMocks }))

describe('services > file', () => {
  test('uploadFile > success', async () => {
    const user = { dn: 'testUser' } as UserDoc
    const modelId = 'testModelId'
    const name = 'testFile'
    const mime = 'text/plain'
    const stream = new Readable() as any
    fileModelMocks.save.mockResolvedValueOnce({ example: true })

    const result = await uploadFile(user, modelId, name, mime, stream)

    expect(s3Mocks.putObjectStream).toBeCalled()
    expect(fileModelMocks.save).toBeCalled()
    expect(result.save).toBe(fileModelMocks.save)
  })

  test('uploadFile > no permission', async () => {
    authorisationMocks.userModelAction.mockResolvedValueOnce(false)

    expect(() => uploadFile({} as any, 'modelId', 'name', 'mime', new Readable() as any)).rejects.toThrowError(
      /^You do not have permission to upload a file to this model./,
    )
  })

  test('removeFile > success', async () => {
    const user = { dn: 'testUser' } as UserDoc
    const modelId = 'testModelId'
    const fileId = 'testFileId'

    const result = await removeFile(user, modelId, fileId)

    expect(result).toMatchSnapshot()
  })

  test('removeFile > no permission', async () => {
    authorisationMocks.userModelAction.mockResolvedValueOnce(false)

    const user = { dn: 'testUser' } as UserDoc
    const modelId = 'testModelId'
    const fileId = 'testFileId'

    expect(() => removeFile(user, modelId, fileId)).rejects.toThrowError(
      /^You do not have permission to delete a file from this model./,
    )

    expect(fileModelMocks.delete).not.toBeCalled()
  })

  test('getFilesByModel > success', async () => {
    const user = { dn: 'testUser' } as UserDoc
    const modelId = 'testModelId'

    const result = await getFilesByModel(user, modelId)

    expect(result).toMatchSnapshot()
  })

  test('getFilesByModel > no permission', async () => {
    authorisationMocks.userModelAction.mockResolvedValueOnce(false)

    const user = { dn: 'testUser' } as UserDoc
    const modelId = 'testModelId'

    expect(() => getFilesByModel(user, modelId)).rejects.toThrowError(/^You do not have permission to get these files./)

    expect(fileModelMocks.delete).not.toBeCalled()
  })
})
