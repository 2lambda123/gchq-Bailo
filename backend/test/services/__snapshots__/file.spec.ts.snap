// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`services > file > getFilesByModel > success 1`] = `
[
  {
    "example": "file",
  },
]
`;

exports[`services > file > removeFile > success 1`] = `
{
  "delete": [MockFunction spy] {
    "calls": [
      [],
    ],
    "results": [
      {
        "type": "return",
        "value": [Circular],
      },
    ],
  },
  "find": [MockFunction spy],
  "findOne": [MockFunction spy] {
    "calls": [
      [
        {
          "_id": "testFileId",
        },
      ],
    ],
    "results": [
      {
        "type": "return",
        "value": [Circular],
      },
    ],
  },
  "save": [MockFunction spy],
  "size": 100,
}
`;
